tests: enable user mode by default

tyrantlizard77 opened the pull request:
Many boards do not enable user mode in their default configuration
even though they support it. However for purposes of our unit
testing we really do want it on all the time.

It's not enough to just flip on CONFIG_USERSPACE, the tests largely
assume that CONFIG_APPLICATION_MEMORY and CONFIG_DYNAMIC_OBJECTS
are enabled. Update CONFIG_TEST_USERSPACE to be on by default if
CONFIG_TEST is active, and hang CONFIG_DYNAMIC_OBJECTS on it too.

This problem was not previously detected as qemu_x86 has the
user mode options in its defconfig and it is currently the only
user mode compatible platform which runs in emulation.

Fixes #7692

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
tests: enable user mode by default

Many boards do not enable user mode in their default configuration
even though they support it. However for purposes of our unit
testing we really do want it on all the time.

It's not enough to just flip on CONFIG_USERSPACE, the tests largely
assume that CONFIG_APPLICATION_MEMORY and CONFIG_DYNAMIC_OBJECTS
are enabled. Update CONFIG_TEST_USERSPACE to be on by default if
CONFIG_TEST is active, and hang CONFIG_DYNAMIC_OBJECTS on it too.

This problem was not previously detected as qemu_x86 has the
user mode options in its defconfig and it is currently the only
user mode compatible platform which runs in emulation.

Fixes #7692

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

andrewboie commented:
This has exposed some more problems.

In frdm_k64f:tests/drivers/build_all/test_build_sensors_a_m / frdm_k64f/tests/drivers/build_all/test_build_sensors_a_m
```
In file included from /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/include/arch/arm/cortex_m/scripts/linker.ld:21:0,
                 from /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/arch/arm/soc/nxp_kinetis/k6x/linker.ld:31:
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/include/linker/linker-defs.h:115:2: error: #error "Max supported kernel objects is 32."
 #error "Max supported kernel objects is 32."
  ^~~~~
```

We need to double this kernel object file limit, 32 is just too small.

in nrf52_pca10040:tests/bluetooth/init/test_controller_dbg:

```
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/subsys/bluetooth/host/smp.c: In function ‘sign_test’:
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/subsys/bluetooth/host/smp.c:3864:1: error: r7 cannot be used in asm here
 }
 ^
zephyr/subsys/bluetooth/host/CMakeFiles/subsys__bluetooth__host.dir/build.make:254: recipe for target 'zephyr/subsys/bluetooth/host/CMakeFiles/subsys__bluetooth__host.dir/smp.c.obj' failed
make[3]: *** [zephyr/subsys/bluetooth/host/CMakeFiles/subsys__bluetooth__host.dir/smp.c.obj] Error 1
```

Here I am truly baffled.

lpereira reviewed (approved):
Seems sane.

Commit:
test_build_sensors_a_m: disable app memory

We run into the limit of 32 object files on ARM when
CONFIG_APPLICATION_MEMORY is enabled.

Bug #7703 filed, meanwhile just disable it, it's not
needed for this test case.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
tests: mem_protect: fix off-by-one

A test was trying to add the maximum number of partitions,
but when the domain was initialized there was already one
added which needed to be accounted for to avoid an
assertion failing.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
elf_helper: fix member offset calculation

On some arches like ARC, the member location tag is a list with
the offset and then the member size. We just need the offset.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
kernel: prepare threads after PRE_KERNEL*

prepare_multithreading() was done very early as it had a call
to initialize the interrupt subsystem. This was causing problems
with stack pointer randomization as any HW-based entropy drivers
had not been initialized.

Move the call to initialize the interrupt system out of
prepare_multithreading(), which now really does just prepare
the system to start threads. This is now done after the PRE_KERNEL
phases.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
tests: fatal: fix several issues

An errant commit accidentally disabled all testing of
hardware-based stack protection. Restore it, and work
around a problem with how these kinds of exceptions are
reported on ARM until #7706 is fixed.

We need to globally disable user mode due to how the
select statements in Kconfig work, the stack sentinel
is incompatible with user mode.

Some build warnings when compiling as native_posix
fixed.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
bluetooth: disable user mode for one test

This is a workaround until #7704 is fixed, we are getting
an assembler error due to the r7 clobber in the syscall
invocation macros.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>
